File: services/interfaces/api/admin/model_config.py
"""Admin API — 模型配置管理路由"""

from __future__ import annotations

from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Query

from application.services.model_config_service import ModelConfigService
from interfaces.api.deps import AdminUser, AppSettings, DbSession
from interfaces.schemas.model_config import (
    ModelConfigCreate,
    ModelConfigDetail,
    ModelConfigOut,
    ModelConfigUpdate,
    ModelProviderCreate,
    ModelProviderDetail,
    ModelProviderOut,
    ModelProviderUpdate,
    SceneModelBindingCreate,
    SceneModelBindingOut,
    SceneModelBindingUpdate,
)
from shared.response import ok, paged

router = APIRouter(
    prefix="/api/v1/admin/model-config",
    tags=["admin-model-config"],
)


def _svc(db, settings) -> ModelConfigService:
    return ModelConfigService(db=db, settings=settings)


# ═══════════════════════════════════════════════════════════════
#  Provider
# ═══════════════════════════════════════════════════════════════


@router.get("/providers", summary="供应商列表")
async def list_providers(
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    is_active: Optional[bool] = None,
):
    svc = _svc(db, settings)
    rows, total = await svc.list_providers(page=page, size=size, is_active=is_active)
    items = [ModelProviderOut.model_validate(r).model_dump() for r in rows]
    return paged(items=items, total=total, page=page, page_size=size)


@router.get("/providers/{provider_id}", summary="供应商详情")
async def get_provider(
    provider_id: UUID,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    row = await svc.get_provider(provider_id)
    return ok(ModelProviderDetail.model_validate(row).model_dump())


@router.post("/providers", summary="新建供应商", status_code=201)
async def create_provider(
    body: ModelProviderCreate,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    row = await svc.create_provider(body)
    await db.commit()
    return ok(ModelProviderOut.model_validate(row).model_dump())


@router.put("/providers/{provider_id}", summary="更新供应商")
async def update_provider(
    provider_id: UUID,
    body: ModelProviderUpdate,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    row = await svc.update_provider(provider_id, body)
    await db.commit()
    return ok(ModelProviderOut.model_validate(row).model_dump())


@router.delete("/providers/{provider_id}", summary="删除供应商")
async def delete_provider(
    provider_id: UUID,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    await svc.delete_provider(provider_id)
    await db.commit()
    return ok()


@router.post("/providers/{provider_id}/test", summary="测试供应商连接")
async def test_provider(
    provider_id: UUID,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    result = await svc.test_provider_connection(provider_id)
    return ok(result)


# ═══════════════════════════════════════════════════════════════
#  Model Config
# ═══════════════════════════════════════════════════════════════


@router.get("/configs", summary="模型配置列表")
async def list_configs(
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    provider_id: Optional[UUID] = None,
    is_active: Optional[bool] = None,
):
    svc = _svc(db, settings)
    rows, total = await svc.list_model_configs(
        page=page, size=size, provider_id=provider_id, is_active=is_active
    )
    items = [ModelConfigOut.model_validate(r).model_dump() for r in rows]
    return paged(items=items, total=total, page=page, page_size=size)


@router.get("/configs/{config_id}", summary="模型配置详情")
async def get_config(
    config_id: UUID,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    row = await svc.get_model_config(config_id)
    return ok(ModelConfigDetail.model_validate(row).model_dump())


@router.post("/configs", summary="新建模型配置", status_code=201)
async def create_config(
    body: ModelConfigCreate,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    row = await svc.create_model_config(body)
    await db.commit()
    return ok(ModelConfigOut.model_validate(row).model_dump())


@router.put("/configs/{config_id}", summary="更新模型配置")
async def update_config(
    config_id: UUID,
    body: ModelConfigUpdate,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    row = await svc.update_model_config(config_id, body)
    await db.commit()
    return ok(ModelConfigOut.model_validate(row).model_dump())


@router.delete("/configs/{config_id}", summary="删除模型配置")
async def delete_config(
    config_id: UUID,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    await svc.delete_model_config(config_id)
    await db.commit()
    return ok()


# ═══════════════════════════════════════════════════════════════
#  Scene Binding
# ═══════════════════════════════════════════════════════════════


@router.get("/bindings", summary="场景绑定列表")
async def list_bindings(
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
    page: int = Query(1, ge=1),
    size: int = Query(50, ge=1, le=200),
):
    svc = _svc(db, settings)
    rows, total = await svc.list_bindings(page=page, size=size)
    items = [SceneModelBindingOut.model_validate(r).model_dump() for r in rows]
    return paged(items=items, total=total, page=page, page_size=size)


@router.post("/bindings", summary="新建场景绑定", status_code=201)
async def create_binding(
    body: SceneModelBindingCreate,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    row = await svc.create_binding(body)
    await db.commit()
    return ok(SceneModelBindingOut.model_validate(row).model_dump())


@router.put("/bindings/{scene_key}", summary="更新场景绑定")
async def update_binding(
    scene_key: str,
    body: SceneModelBindingUpdate,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    row = await svc.update_binding(scene_key, body)
    await db.commit()
    return ok(SceneModelBindingOut.model_validate(row).model_dump())


@router.delete("/bindings/{scene_key}", summary="删除场景绑定")
async def delete_binding(
    scene_key: str,
    db: DbSession,
    settings: AppSettings,
    _user: AdminUser,
):
    svc = _svc(db, settings)
    await svc.delete_binding(scene_key)
    await db.commit()
    return ok()

File: admin/src/api/model-config.ts
import { client, unwrap } from './client';

// ── Types ──

export interface ModelProvider {
  id: string;
  name: string;
  provider_type: string;
  base_url: string | null;
  api_key_masked: string;
  extra_config: Record<string, unknown> | null;
  is_active: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface ModelConfig {
  id: string;
  provider_id: string;
  model_name: string;
  display_name: string;
  capabilities: string[];
  default_params: Record<string, unknown>;
  is_active: boolean;
  sort_order: number;
  created_at: string;
  updated_at: string;
}

export interface ModelConfigDetail extends ModelConfig {
  provider: ModelProvider | null;
}

export interface SceneModelBinding {
  id: string;
  scene_key: string;
  scene_label: string | null;
  model_config_id: string;
  param_overrides: Record<string, unknown>;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

// ── Providers ──

export function listProviders(params: { page?: number; size?: number } = {}) {
  return client.get('/admin/model-config/providers', { params }).then(unwrap) as Promise<{ items: ModelProvider[]; total: number }>;
}

export function getProvider(id: string) {
  return client.get(`/admin/model-config/providers/${id}`).then(unwrap) as Promise<ModelProvider>;
}

export function createProvider(data: {
  name: string;
  provider_type: string;
  base_url?: string;
  api_key: string;
  extra_config?: Record<string, unknown>;
  is_active?: boolean;
  sort_order?: number;
}) {
  return client.post('/admin/model-config/providers', data).then(unwrap) as Promise<ModelProvider>;
}

export function updateProvider(id: string, data: Partial<{
  name: string;
  provider_type: string;
  base_url: string;
  api_key: string;
  extra_config: Record<string, unknown>;
  is_active: boolean;
  sort_order: number;
}>) {
  return client.put(`/admin/model-config/providers/${id}`, data).then(unwrap) as Promise<ModelProvider>;
}

export function deleteProvider(id: string) {
  return client.delete(`/admin/model-config/providers/${id}`);
}

// ── Model Configs ──

export function listModelConfigs(params: { provider_id?: string; page?: number; size?: number } = {}) {
  return client.get('/admin/model-config/configs', { params }).then(unwrap) as Promise<{ items: ModelConfig[]; total: number }>;
}

export function getModelConfig(id: string) {
  return client.get(`/admin/model-config/configs/${id}`).then(unwrap) as Promise<ModelConfigDetail>;
}

export function createModelConfig(data: {
  provider_id: string;
  model_name: string;
  display_name: string;
  capabilities?: string[];
  default_params?: Record<string, unknown>;
  is_active?: boolean;
  sort_order?: number;
}) {
  return client.post('/admin/model-config/configs', data).then(unwrap) as Promise<ModelConfig>;
}

export function updateModelConfig(id: string, data: Partial<{
  provider_id: string;
  model_name: string;
  display_name: string;
  capabilities: string[];
  default_params: Record<string, unknown>;
  is_active: boolean;
  sort_order: number;
}>) {
  return client.put(`/admin/model-config/configs/${id}`, data).then(unwrap) as Promise<ModelConfig>;
}

export function deleteModelConfig(id: string) {
  return client.delete(`/admin/model-config/configs/${id}`);
}

// ── Scene Bindings ──

export function listSceneBindings(params: { page?: number; size?: number } = {}) {
  return client.get('/admin/model-config/bindings', { params }).then(unwrap) as Promise<{ items: SceneModelBinding[]; total: number }>;
}

export function createSceneBinding(data: {
  scene_key: string;
  scene_label?: string;
  model_config_id: string;
  param_overrides?: Record<string, unknown>;
  is_active?: boolean;
}) {
  return client.post('/admin/model-config/bindings', data).then(unwrap) as Promise<SceneModelBinding>;
}

export function updateSceneBinding(sceneKey: string, data: Partial<{
  scene_label: string;
  model_config_id: string;
  param_overrides: Record<string, unknown>;
  is_active: boolean;
}>) {
  return client.put(`/admin/model-config/bindings/${sceneKey}`, data).then(unwrap) as Promise<SceneModelBinding>;
}

export function deleteSceneBinding(sceneKey: string) {
  return client.delete(`/admin/model-config/bindings/${sceneKey}`);
}
